### PR TITLE
ledger-tool: Remove duplicate slot output code

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
Currently, `ledger-tool slot` has some duplicate implementation for outputting a block. Also, the `--output json` mode for the `slot` command is effectively broken / uselss.

#### Summary of Changes
This PR removes some of the duplicate implementation, and makes the outputs from `ledger-tool blockstore slot -vv` and `ledger-tool bigtable block` follow the same codepath. The bigtable command with `--show-entries` flag will result in identical output assuming the blockstore has the metadata fields (ie node was running with `--enable-rpc-transaction-history`)

### CODE REVIEW

File: ledger-tool/src/output.rs [hunk 1]
```diff
@@ -327,162 +331,116 @@ impl EncodedConfirmedBlockWithEntries {
...
     }
 
-    let (entries, num_shreds, is_full) = blockstore
-        .get_slot_entries_with_shred_info(slot, 0, allow_dead_slots)
-        .map_err(|err| format!("Failed to load entries for slot {slot}: {err:?}"))?;
+    let Some(meta) = blockstore.meta(slot)? else {
+        return Ok(());
+    };
+    let VersionedConfirmedBlockWithEntries { block, entries } = blockstore
+        .get_complete_block_with_entries(
```
Comment (steviez) [inline ledger-tool/src/output.rs:375]: Use of this function now requires a full block to be present, we would previously make a best effort to display entries based on what we had available. I can see how displaying partial slots could be good, but thinking personally on this, I usually have two cases:
- See if I have the full slot
- Print the full slot

So, I have convinced myself that only printing full blocks here is ok, but curious for a second opinion. There is probably some common logic that could be factored into a helper

File: ledger-tool/src/output.rs [hunk 2]
```diff
@@ -327,162 +331,114 @@ impl EncodedConfirmedBlockWithEntries {
...
-                    entries.len(),
-                    is_full,
-                );
-            }
+    if verbose_level == 0 {
+        if *output_format == OutputFormat::Display {
+            println!(
+                "  num_shreds: {}, parent_slot: {:?}, next_slots: {:?}, num_entries: {}, \
+                 is_full: {}",
+                meta.consumed,
```
Comment (steviez) [inline ledger-tool/src/output.rs:389]: We already know we have a full block, so `meta.consumed` is the number of shreds. This is of course contingent on us being ok with full-block only requirement.

But, as I review this myself, this maybe isn't super obvious. I can add a comment if this also seems non-obvious to you

**Review by steviez — COMMENTED** [review]
@CriesofCarrots - I took a stab at this, CI hasn't finished yet but there are some open items in my mind. Two of them I called out in code, and a few more below:
- The printing of program usage for each slot is kind of thrown in there. I think this should probably be turned on via a flag or something, but I left it alone for now
- I adjusted the layout of the switch-on-verbosity to make it more clear what is output at each level. Seeing it now, I'm wondering if a new command `slot-meta` might be appropriate for what is verbosity level 1. We could print the `SlotMeta` as well as the accompanying `Index` (the sets that tell us which all shred indices are present)
- Using `--output json` with verbosity `0` or `1` will now output nothing. I think we could easily print json output that matches the display output for these levels, but held off on that given my proposal for a new command

File: ledger-tool/src/output.rs (hunk 1 above)
Comment (CriesofCarrots) [reply ledger-tool/src/output.rs:375]: We return `SlotUnavailable` for a partial slot, yeah? That's less explicit than it could be. What do you think about emitting a trace or debug log if `get_complete_block_with_entries` is called on a partial slot?

It might be nice to have some way to print a partial slot, just in case, but not sure what that UI should be. I'm fine with normal `agave-ledger-tool slot` operation printing only full slots.

**Review by CriesofCarrots — DISMISSED** [review]
Nice!

File: ledger-tool/src/output.rs (hunk 1 above)
Comment (steviez) [reply ledger-tool/src/output.rs:375]: > What do you think about emitting a trace or debug log if get_complete_block_with_entries is called on a partial slot?

Works for me - I see another `info` that is arguably a bit noisy, going to demote that one to trace as well while I'm in there. I force pushed to resolve a conflict, but this is done in new commit https://github.com/anza-xyz/agave/pull/1255/commits/504fd2d7268baf9d951aaeaeed42b09cac79eb5c

> It might be nice to have some way to print a partial slot, just in case, but not sure what that UI should be.

Yeah, I'll think on this a bit more

> I'm fine with normal agave-ledger-tool slot operation printing only full slots.

👍

File: ledger-tool/src/output.rs (hunk 2 above)
Comment (steviez) [reply ledger-tool/src/output.rs:389]: Ended up adding a quick comment in https://github.com/anza-xyz/agave/pull/1255/commits/025ed799e05ffc36a5a52e84c3081c214170d82c

File: ledger-tool/tests/basic.rs [hunk 3]
```diff
@@ -1,12 +1,10 @@
 use {
     assert_cmd::prelude::*,
-    solana_entry::entry,
     solana_ledger::{
         blockstore, blockstore::Blockstore, blockstore_options::ShredStorageType,
-        create_new_tmp_ledger, create_new_tmp_ledger_fifo, genesis_utils::create_genesis_config,
-        get_tmp_ledger_path_auto_delete,
+        create_new_tmp_ledger_auto_delete, create_new_tmp_ledger_fifo_auto_delete,
+        genesis_utils::create_genesis_config, get_tmp_ledger_path_auto_delete,
```
Comment (steviez) [inline ledger-tool/tests/basic.rs:6]: I was already in here so decided to swap - put it in a different commit + low line count so hopefully not too distracting

File: ledger-tool/tests/basic.rs [hunk 4]
```diff
@@ -37,66 +31,49 @@ fn bad_arguments() {
...
-    // Print everything
-    let output = run_ledger_tool(&["-l", ledger_path, "print", "-vvv"]);
+    let output = run_ledger_tool(&["-l", ledger_path, "print", "-vv"]);
     assert!(output.status.success());
-    assert_eq!(
-        count_newlines(&output.stdout)
-            .saturating_sub(meta_lines)
-            .saturating_sub(summary_lines),
-        ticks
-    );
```
Comment (steviez) [inline ledger-tool/tests/basic.rs:55]: I don't think we need to be counting the number of lines of output from the command; we don't claim stability for this kind of output so I think confirming success is sufficient

File: ledger-tool/tests/basic.rs [hunk 5]
```diff
@@ -37,66 +31,49 @@ fn bad_arguments() {
...
 fn insert_test_shreds(ledger_path: &Path, ending_slot: u64) {
     let blockstore = Blockstore::open(ledger_path).unwrap();
-    for i in 1..ending_slot {
-        let entries = entry::create_ticks(1, 0, Hash::default());
-        let shreds = blockstore::entries_to_test_shreds(
-            &entries, i, 0, false, 0, /*merkle_variant:*/ true,
-        );
-        blockstore.insert_shreds(shreds, None, false).unwrap();
-    }
+    let (shreds, _) = blockstore::make_many_slot_entries(
```
Comment (steviez) [inline ledger-tool/tests/basic.rs:58]: The old method was creating partial slots, so `slot` was returning error status code. I switched over to a different helper that is less involved AND also creates a full block of ticks

File: ledger-tool/src/output.rs [hunk 6]
```diff
@@ -325,162 +329,116 @@ impl EncodedConfirmedBlockWithEntries {
...
-                    num_shreds,
-                    meta.parent_slot,
-                    meta.next_slots,
-                    entries.len(),
-                    is_full,
-                );
-            }
+    if verbose_level == 0 {
+        if *output_format == OutputFormat::Display {
+            // Given that Blockstore::get_complete_block_with_entries() returning Ok(_), we know
```
Comment (CriesofCarrots) [inline ledger-tool/src/output.rs]:
```suggestion
            // Given that Blockstore::get_complete_block_with_entries() returned Ok(_), we know
```
grammar nit

File: ledger/src/blockstore.rs [hunk 7]
```diff
@@ -2574,23 +2579,49 @@ impl Blockstore {
...
+    fn do_get_complete_block_with_entries(
+        &self,
+        slot: Slot,
+        require_previous_blockhash: bool,
+        populate_entries: bool,
+        allow_dead_slots: bool,
     ) -> Result<VersionedConfirmedBlockWithEntries> {
         let Some(slot_meta) = self.meta_cf.get(slot)? else {
-            info!("SlotMeta not found for slot {}", slot);
+            trace!("do_get_complete_block_with_entries() failing for {slot} (missing SlotMeta)");
```
Comment (CriesofCarrots) [inline ledger/src/blockstore.rs]:
```suggestion
            trace!("do_get_complete_block_with_entries() failed for {slot} (missing SlotMeta)");
```

File: ledger/src/blockstore.rs [hunk 8]
```diff
@@ -2670,6 +2706,7 @@ impl Blockstore {
                 return Ok(VersionedConfirmedBlockWithEntries { block, entries });
             }
         }
+        trace!("do_get_complete_block_with_entries() failing for {slot} (slot not full)");
```
Comment (CriesofCarrots) [inline ledger/src/blockstore.rs]:
```suggestion
        trace!("do_get_complete_block_with_entries() failed for {slot} (slot not full)");
```

**Review by CriesofCarrots — DISMISSED** [review]
Just message grammar nits, take 'em or leave 'em. Functionality looks great!

**Review by CriesofCarrots — APPROVED** [review]

**Comment by steviez** [comment]
> Just message grammar nits, take 'em or leave 'em. Functionality looks great!

Good pickups, they read better now
